src/nvme/mi.c

```diff
@@ -692,8 +692,10 @@ int nvme_mi_admin_xfer(nvme_mi_ctrl_t ctrl,
 
 	/* limit the response size, specify offset */
 	admin_req->flags = 0x3;
-	admin_req->dlen = cpu_to_le32(resp.data_len & 0xffffffff);
-	admin_req->doff = cpu_to_le32(resp_data_offset & 0xffffffff);
+	admin_req->dlen = req_data_size ? req_data_size: 
+			cpu_to_le32(resp.data_len & 0xffffffff);
+	admin_req->doff = req_data_size ? 0: 
+			cpu_to_le32(resp_data_offset & 0xffffffff);
```

Comment:
You're missing the endian conversion in cases where `req_data_size` is set (ie, the case you care about). Given the logic here is getting more complex, I would suggesting separating this out a little - calculate the values first, then do the endian conversion when setting up the `admin_req` struct. 

Something like:


```c
__u32 dlen, doff;

[...]

/* dlen and doff have different interpretations depending on the data direction */
if (req_data_size) {
    dlen = req_data_size & 0xffffffff;
    doff = 0;
} else {
    dlen = *resp_data_size & 0xffffffff;
    doff = resp_data_offset & 0xffffffff;
}
admin_req->dlen = cpu_to_le32(dlen);
admin_req->doff = cpu_to_le32(doff);
```

We do already know that `resp_data_offset` will always 0 for the former case (due to the checks above), but no harm in being explicit about this.